src/modules/edc-demo/components/asset-viewer/asset-viewer.component.html

```diff
@@ -33,13 +33,28 @@
 
     <!-- Paginator -->
     <mat-paginator
+      *ngIf="assetList.isReady"
       hidePageSize="true"
       pageSize="contractDefinitions.length"
-      [length]="assets.length">
+      [length]="assetList.data.numTotalAssets">
     </mat-paginator>
   </div>
 
+  <div class="flex">
+    <mat-progress-spinner
+      *ngIf="assetList.isLoading"
+      style="width: 80px; height: 80px"
+      diameter="80"
+      color="primary"
+      mode="indeterminate">
+    </mat-progress-spinner>
+    <div *ngIf="assetList.isError" class="text-center mt-[50px] w-full">
+      <span>{{ assetList.error.failureMessage }}</span>
+    </div>
+
   <edc-demo-asset-list
-    [assets]="assets"
+    *ngIf="assetList.isReady && assetList.data.numTotalAssets >= 1"
+    [assets]="assetList.data.filteredAssets"
     (assetClick)="onAssetClick($event)"></edc-demo-asset-list>
+  </div>
```

Comment:
Please create re-usable components "empty-state", "loading-state" and "error state" that will be used to display full page / emptyness / loading / errors.

empty state should have as parameter the empty message

loading state needs to parameters unless you find a loading message nice to use

error state should be passed the error object from fetched, where you have both the error and the failure message


Check out how dashboard-card works out. We want these re-usable components to center the content in the HTML tag of the component. For that we need to add classes to the Component HTML Tag. You can see how to do that in `dashboard-donut-chart.component.ts` in the section with @HostBinding.